controller/src/main/java/io/pravega/controller/eventProcessor/impl/EventProcessorGroupImpl.java

```diff
@@ -214,64 +252,50 @@ protected void shutDown() {
-            cell.stopAsync();
-        }
-        
-        for (EventProcessorCell<T> cell : eventProcessorMap.values()) {
-            try {
-                cell.awaitTerminated();
-                log.debug("Termination of event processor cell: {} completed successfully.", cell);
-            } catch (Exception e) {
-                log.warn("Failed terminating event processor cell {}.", cell, e);
-            }
-        }
-        
-        // Finally, clean up reader group from checkpoint store.
+    @Override
+    public void notifyProcessFailure(String controllerId) throws CheckpointStoreException {
+        long traceId = LoggerHelpers.traceEnterWithContext(log, this.objectId, "notifyProcessFailure", controllerId);
+        log.info("Notifying failure of process {} participating in reader group {}", controllerId, this.objectId);
         try {
-            log.debug("Attempting to clean up reader group {} entry from checkpoint store", this.objectId);
-            checkpointStore.removeProcessFromGroup(actorSystem.getProcess(), readerGroup.getGroupName());
-        } catch (CheckpointStoreException e) {
-            log.warn("Error removing reader group " + this.objectId, e);
-            return;
+            Map<String, Position> map = sealReaderGroup(controllerId);
+            cleanupReaderGroup(controllerId, map);
+        } finally {
+            LoggerHelpers.traceLeave(log, "notifyProcessFailure", traceId, controllerId);
         }
     }
 
-    @Override
-    public void notifyProcessFailure(String process) throws CheckpointStoreException {
-        long traceId = LoggerHelpers.traceEnterWithContext(log, this.objectId, "notifyProcessFailure", process);
-        log.info("Notifying failure of process {} participating in reader group {}", process, this.objectId);
+    private Map<String, Position> sealReaderGroup(String controllerId) throws CheckpointStoreException {
+        Map<String, Position> readerPositions = ImmutableMap.of();
         try {
-            Map<String, Position> map;
-            try {
-                log.debug("Removing reader group {} from process {}", readerGroup.getGroupName(), process);
-                map = checkpointStore.removeProcessFromGroup(process, readerGroup.getGroupName());
-            } catch (CheckpointStoreException e) {
-                if (e.getType().equals(CheckpointStoreException.Type.NoNode)) {
-                    return;
-                } else {
-                    throw e;
-                }
+            log.debug("Attempting to seal Reader Group {} for Controller {}", readerGroup.getGroupName(), controllerId);
+            // Change Reader Group State to SEALED so no new Readers can join
+            readerPositions = checkpointStore.sealReaderGroup(controllerId, readerGroup.getGroupName());
+        } catch (CheckpointStoreException e) {
+            if (!e.getType().equals(CheckpointStoreException.Type.NoNode)) {
```

Comment:
nit: This act of suppressing the NoNode exception can be done inside the ZKCheckpointStore and we could eliminate this method from EventProcessorGroupImpl.

Reply:
That does not feel like the right thing to do. For a remove operation, if the node is already removed it should be idempotent so there is no benefit in throwing the NoNode Exception, but that cannot be said about sealReaderGroup. I'd want the checkpoint store method to throw an exception and we handle it here, cause here we know its fine to not be able to find that node....